### ISSUE TITLE
Stash unstaged changes prior to start

### ISSUE THREAD

**Issue by Zoramite** [nvie/gitflow#32]
Currently when starting a new feature/etc on a 'dirty' working directory it shows a message:

```
fatal: Working tree contains unstaged changes. Aborting.
```

What if we changed that functionality do to a <code>git stash</code> of the before creating the branch, create the new branch for the feature/etc, then <code>git stash apply</code> after switching to the new branch?

**Comment by nvie** [comment]
Fair. I might have been a bit too strict with this one. (It regularly itches me, too.)

**Comment by nvie** [comment]
Maybe not stash the changes away, but simply remove the check for local changes and let git determine if it can switch branches itself?

**Comment by Zoramite** [comment]
Looking into the specific commands it looks like this basic idea would work:

```
git stash
... do stuff ...
git stash pop
```

if it dies while 'doing stuff' we may want to do a <code>git stash pop</code> before exiting to get it back to where it was before it was started.

**Comment by Zoramite** [comment]
That is true, git will give its own error if it can't branch.

**Comment by nvie** [comment]
I'm tempted to have a very conservative additude against these kind of implementations. Think worst case: what if `git stash` fails (somehow)? And what if `git stash pop` fails? That's two more sanity checks and remembering state in between. Not to mention that Git's porcelain does not provide the best ways to do these sanity checks in the first place.

If this can be tackled by _removing_ one of our sanity checks and leaving the figuring out of whether branching is currently possible to Git, we have a solid implementation. If Git is unable to start a new branch somehow, we should quit and let the user deal with it. This is probably for the better anyway, since that should be an exceptional case.

**Comment by Zoramite** [comment]
Agreed, it would be difficult for gitflow to maintain the state. I'm all for just removing the extra check.

**Comment by nvie** [comment]
Heh, this is a bit lame: looking at the source code, I apparently added the option `-f` (force) for this already. The question now is: should this become the default behaviour or just keep it this way?

**Comment by Zoramite** [comment]
From my ignorant view it would seem better to have it try the branch and fail with the git message than have to fail, then force, then fail again.

But we may need to look into what other actions occur between when the check for a 'dirty' workspace and the branch code. Is there anything that gets stored between the check and the branch creation that would mess up the process?

**Comment by nvie** [comment]
Yeah, you're absolutely right. Fixed in 4132de9ec4c94d6de3282c9b8f8cb55675a24769.
